add citation, update readthedoc config, add 3.12 in CI

rcaneill opened the pull request:
Closes #66

rcaneill commented:
https://support.zenodo.org/help/en-gb/24-github-integration/96-how-does-a-citation-cff-file-affect-metadata-of-my-github-release

For our needs, the CITATION.cff file seems more relevant that the .zenodo.json file

rcaneill commented:
@DocOtak I let you add your email address if you wish so

rcaneill commented:
https://docs.readthedocs.io/en/stable/build-customization.html#install-dependencies-with-poetry

rcaneill commented:
The python 3.12 error seems to come from:
1. the fact the numpy version that supports 3.12 dropped 3.8
2. jupyterlab needs to be updated

DocOtak commented:
Looks like xarray also dropped 3.8 support, we should do the same.

Some versions policies: [SPEC0](https://scientific-python.org/specs/spec-0000/), [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html#support-table), Xarray [follows NEP 29 for now](https://docs.xarray.dev/en/stable/getting-started-guide/installing.html#minimum-dependency-versions) but I know they discuss switching to SEPC0. If I recall correctly, SPEC0 is faster to drop support for things.

I personally usually keep my python one point release behind, so when 3.12 came out, I migrated my personal work from 3.10 to 3.11.

rcaneill commented:
Seems ready to merge for me.
I dropped python 3.8 and added python 3.12.
I added a test to verify that the minimal version available in the pyproject file are passing the tests (and I discovered that we need pint>=0.20)

rcaneill commented:
Seems ready to merge for me.
I dropped python 3.8 and added python 3.12.
I added a test to verify that the minimal version available in the pyproject file are passing the tests (and I discovered that we need pint>=0.20)

rcaneill commented:
@DocOtak could you review?

DocOtak commented:
Nothing jumped out at me reading the changes with my eyes, which is the classic "looks good to me" type response I suppose. I'll poke at the actual code tomorrow morning (UTC-7). It might be worth experimenting with switching things as much as possible to ruff in a later PR. Absolutely support you being "first author" on anything related to this if the ordering in that citation.cff file matters.

---

P.S. I did see your PhD advisor at Ocean Sciences, he was quite busy with folks interested in his poster so I didn't actually get the opportunity to introduce myself.

DocOtak commented on .github/workflows/lint.yml:
Does this action work? I only see failures from 6mo ago in the action logs. There are a lot of deprecated things in this as well I think like the setup-python action, which I think the most recent is v4 or v5. There is also an old version of python being specified.

rcaneill replied:
Well caught

DocOtak commented on poetry.lock:
We might consider dropping the lock file since we are developing a library: https://python-poetry.org/docs/basic-usage/#as-a-library-developer unless it's needed for CI.

rcaneill replied:
I think that we can keep it as it make the CI run faster

rcaneill commented:
@DocOtak I moved everything to ruff instead of fixing stuff that we would have removed later. Can you review?

DocOtak commented:
Awesome, I think this is ready to go in. I guess there hasn't been a black run for a while as the style is slightly changing. And I don't know if isort as ever been run on the project.

This should close #70 since flake8 will basically be dropped.
